### ISSUE TITLE
Provided accumulators should support fact retraction

### ISSUE THREAD

**Issue by rbrush** [rbrush/clara-rules#127]
The accumulators in clara.rules.accumulators namespace don't handle handle retraction of facts, which can lead to confusing behavior. For instance, the count accumulator won't decrement the count if one of the items it is counting is retracted from the working memory. This and others (such as c.r.a/sum, c.r.a/average, c.r.a/all, c.r.a/distinct) can be easily added by providing a retract-fn.

In the example of the count accumulator, the current code is:

```clj
(defn count
  "Returns an accumulator that simply counts the number of matching facts"
  []
  (accumulate
   :initial-value 0
   :reduce-fn (fn [count value] (inc count))
   :combine-fn +))
```

We simply need to add a retract-fn like this:

```clj
   :retract-fn (fn [count retracted] (dec count))
```

**Comment by devn** [comment]
Maybe I'm off base here, but if this is a sensible default, then perhaps it
wouldn't be a bad idea to make the definition of an accumulator enforce the
presence of a retract fn.

**Comment by mrrodriguez** [comment]
I think this issue strongly related to
https://github.com/rbrush/clara-rules/issues/86

Which I don't think we ever really came to a resolution on. 
Perhaps reading that will give you some background devn.

**Comment by mrrodriguez** [comment]
For what it's worth though. The way we dealt with this for our usage is we provided a general purpose "wrapper" for creating Clara accumulators. It turns he default "non-state-preserving" accumulators into an accumulator that implicitly keeps track of the state of everything seen so far and a default, naiive retract fn that just runs an entire reduce over all facts seen, minus what has been retracted. This is less than ideal but maintains consistency with the truth maintenance of the engine and performs alright if there aren't heavy retraction rework happening. 
The engine may be able to provide a better default from within the engine itself perhaps with something like batches retracts instead of one at a time. However I haven't gotten to think about this much yet. 

The issue I linked to above is what the inspiration of the general purpose wrapper came from though.

**Comment by rbrush** [comment]
@mrrodriguez, thanks for bringing up #86. Looking at it now, I think the default behavior I mentioned in the discussion in #86 wasn't the right choice. I think Clara's provided accumulators should behave in the least surprising way possible, even if that's not as efficient as an accumulator that doesn't have to deal with retractions. To @devn's point, I agree that requiring the caller to specify a retract-fn also makes sense...and if a caller wants to define an accumulator that doesn't keep old state he is welcome to do so and have a retract-fn that does nothing. (This change could break users that don't offer one, but it will fail fast with a clear message and I think the 0.9.0 stream is a good point to introduce such changes.)

The nice thing is that Clara's accumulator abstraction makes it easy for users to define their own accumulators that do whatever makes sense for the case at hand. So @mrrodriguez's wrapper model could continue to be used there.

**Comment by rbrush** [comment]
I went ahead and added retract support and tests to the provided accumulators in the above commit. I'll probably merge it to master soon but wanted it in the branch in case anyone wanted to take a look or suggest a different course beforehand.

**Comment by mrrodriguez** [comment]
I think this approach is interesting and will work with minimal effort.

I want to discuss a bit on the tradeoffs of this approach though:

---

This will mean that the `:reduce-fn` for many accumulators will in a way stop behaving like the intuitive notion of what a reducing function means in a typical `reduce` operation. 
e.g. If I'm writing an accumulator to find the newest timestamp on some fact, I'd expect to write the comparison logic for the newest in the `:reduce-fn`.  With this approach, the `:reduce-fn` often does little more than collect everything to delay the "reducing decision/step" until the `:convert-return-fn`.

To summarize:
* This is a loss of intuition around what the `:reduce-fn` means (to me and perhaps others).
* This is, in a way, a leak of the underlying engine details.  The consumer is expected to manage their own accumulator memory.
* This isn't how Drools works with accumulators.  It will just rerun the accumulate with the item removed if no "reverse" operation is provided.  I'm not sure about other mainstream engines.  This may not be that important though.
* If the `:convert-return-fn` becomes heavy-weight I do worry that this logic isn't happening in the "ideal place" since it isn't even in the the accumulators `reduce`, but rather after it.  That may just be a misguided performance concern though mostly.  I think we would be losing the ability for parallel processing the reducing work though (which isn't used right now, but sort of positioned to be).

--

I think I mostly understand the tradeoff here.  This approach is simple and allows the engine to continue to not maintain any memory within these nodes.  

However, what are we gaining from the engine not maintaining this node memory though?  There is a certain "stream" ability for an accumulate node right now, but does that matter when we have to retain all of these references in (global) working memory anyways?  

I think this may relate some to the distributed Storm approach that I have not had a lot of time to work with at this point.  Couldn't we just add a flag to an accumulator definition if we wanted it to behave in a streaming way, so that that became the edge case rather than changing the typical all-in-memory case?

--

If the engine did maintain accumulate memory, we could have a naive `:retract-fn` for "free" that just did what this commit calls `clara.rules.accumulator/drop-one-off` and then performed a `reduce` with the reducing function being the `:retract-fn` (Drools does similar).

---

Given my points above, I'm still leaning towards preferring the Clara engine maintaing accumulate node memory.  I am curious for your feedback on this though, especially around what tradeoffs we are working with precisely.

**Comment by rbrush** [comment]
We could easily change the min and max accumulators to maintain a heap of values in the reduce-fn, and the convert-return-fn would simply pull the item off the heap. But lacking a convenient heap structure I suspect it's more efficient to do the sort at return time, rather than doing a sort as values are added to the reduce-fn. I don't think this pattern of do-the-work-when-we-return-the-value is necessarily representative of other accumulators...it was just the simplest implementation that presented itself.

I would be okay with adding a :supports-retract option (and default it to true), so an implementation of min/max could keep only the actual value rather than all records. It looks like this is actually what Drools does by default.

Managing this state efficiently when hosting rules in a distributed system is a pretty complicated topic in itself. If we provide the option to let things work as they do now, users of a distributed system could use that to gain efficiency.  We should consider ways to support richer semantics when distributed as well, but I'm not quite ready to get into that here. ;)

**Comment by mrrodriguez** [comment]
I agree that the implementation choice for min and max probably are doing it in an efficient way.  Perhaps that would be the case for many accumulators and my points off-base.

Maybe Clara should just provide an accumulator wrapper impl that would turn any accumulator into one that had a `:retract-fn` that would be able to retract automatically.  This requires managing some "hidden state" of the facts seen behind the `:initial-value`, `:reduce-fn`, and `:combine-fn`, but isn't too hard to do.

The wrapper approach could be sufficient if it isn't going to be any slower than what the engine would be capable of doing with specific node memory (I haven't gotten to think about that one enough yet).

The only drawback to this to me still is that without the wrapper the default behavior is surprising.  
To reduce the surprise, we could keep the `:retract-fn` as required and you could just call the decorator when you wanted one for "free".
Something like:
```clj
(clara.rules/accumulate (clara.rules/with-retract-support {:reduce-fn my-reduce-fn}))
```

**Comment by rbrush** [comment]
I'm a bit reluctant to add the wrapper pattern into Clara itself since I'm not sure it's a common, general pattern. We would definitely support that pattern, but it might make more sense for it to live in user libraries. I think this is still open for debate...I just want to keep the API small.

For users wanting to create accumulators with simple reduce functions, I think we can add some functions for that as a convenience. What if we deprecate today's "accumulate" function add add an overloaded "accumulator" function like this:

```cli
(accumulator {:initial-value 0, :reduce-fn +, combine-fn + retract-fn -}) ; Creates an accumulator with all options available today.
(accumulator my-reduce-fn) ; Creates a simple accumulator that uses the given reduce method
(accumulator my-reduce-fn :update-on-retraction false) ; Creates an accumulator without retraction support.
```

Basically the ```accumulator``` function would accept either a map or a simple reduce function and options to grant more control for special use cases, and everything would have defaults resulting in the least surprising behavior. I think we'd deprecate clara.rules/accumulate and replace it with this (perhaps moving it to the clara.rules.accumulators namespace since it seems to fit better there.)

**Comment by mrrodriguez** [comment]
+1
I like your last proposal here.

Are you thinking
```clj
(accumulator my-reduce-fn)
``` 
has `:update-on-retraction` defaulting to true and that means that Clara will provide the retraction capabilities to the accumulator?  Or if it isn't the default, is that what true means still?

**Comment by rbrush** [comment]
Yes, :update-on-retraction would default to true and Clara will provide the retraction capabilities itself.

**Comment by rbrush** [comment]
Alright, I got a chance to get back to this today. I ended up going with new accum and reduce-to-accum functions in the accumulators namespace, and patterned it after some logic Mike had worked out to convert reduce functions into accumulators. 

All built-in accumulators now support retraction properly by default, and it should be easier to create user-defined accumulators that do so as well. Feedback is welcome. If nothing comes up I'll merge to master in the next few days.

**Comment by mrrodriguez** [comment]
I think this commit looks good. Thanks for all the rounds of quick feedback on this issue too!

**Comment by rbrush** [comment]
Merged! Thanks for the input.

**Comment by devn** [comment]
:clap: Thanks to you all for your work on this.
